### ISSUE TITLE
Implement a consistent error parsing system for the client

### ISSUE THREAD

**Issue by joshuagraber** [Police-Data-Accessibility-Project/pdap.io#199]
### Context

Brief description. Is this part of a parent issue?

### Requirements

* Parse errors delivered from the API in a human-readable, sensible format.

### Tests

* N/A

### Docs

* N/A

### Open questions

* @joshuagraber will audit the errors thrown by the API, map them to informative errors that can be delivered in the UI, with input from @maxachis and @josh-chamberlain

**Comment by joshuagraber** [comment]
Adding hold label, let's wait on this until @joshuagraber completes the audit of the error messages thrown by the API

**Comment by maxachis** [comment]
@joshuagraber Where are we at with this?

**Comment by joshuagraber** [comment]
@maxachis this work has not been started.

**Comment by maxachis** [comment]
Here's the Claude Code audit on this!

## Audit & proposed architecture for client error parsing

Posting a survey of current error handling in the client + a proposed shape for the parsing system, so @joshuagraber's upcoming API-error audit has a concrete target to feed into.

### Current state

Stack: Vue 3 + Vite, `axios` for HTTP, `vue3-toastify` for notifications, Vuelidate for form validation, TanStack Vue Query for most mutations/queries.

**No central HTTP client.** API modules in `src/api/*` call `axios` directly with no shared instance, no interceptors, and no typed error responses. Examples: `src/api/auth.js:16`, `src/api/search.js:23`, `src/api/annotate.ts:29`.

**Error handling is inconsistent — at least three patterns coexist:**
1. Silent/console-only — `src/api/auth.js:108`, `src/api/typeahead.js:22`
2. Ad-hoc toasts — `src/pages/request-reset-password.vue:89`, `src/pages/profile/index.vue:352`
3. Inline form errors via Vuelidate — `src/pages/data-requests/create.vue:431`, `src/pages/data-sources/create.vue:792`

**Error shapes consumed by the UI vary:**
- `error.response.data.message` — `src/pages/sign-in.vue:223`, `src/pages/sign-up/index.vue:271`
- `err.message` fallback — `src/pages/request-reset-password.vue:90`
- `error instanceof Error` branching — `src/pages/profile/index.vue:375`
- String matching on message contents — sign-in/sign-up

No shared error types, no parsing utility, and no convention for *which* surface (toast vs. inline vs. page) an error belongs on.

### Proposed architecture

Four layers, each thin and replaceable:

1. **Typed error response contract** (`src/api/errors/types.ts`) — a discriminated union describing the shapes the API actually returns (populated from @joshuagraber's audit). Include an `UnknownApiError` fallback so the system is safe before the audit is complete.

2. **Central axios instance + response interceptor** (`src/api/client.ts`) — replaces direct `axios` imports. The interceptor normalizes any caught error into a single `AppError { code, userMessage, field?, severity, cause }` object, so downstream code never touches `error.response.data` directly.

3. **Error-message map** (`src/api/errors/messages.ts`) — the output of the audit: `code → userMessage`, with i18n-friendly keys. Unknown codes fall back to a generic message + log the raw error.

4. **Presentation composable** (`src/composables/useErrorHandler.ts`) — decides *where* an `AppError` surfaces: field-scoped errors route to the form, others to a toast. Integrates cleanly with Vue Query's `onError` and replaces the scattered toast calls.

### Rollout

- Land layers 1, 2, 4 with a passthrough/generic message map — no behavior change.
- Migrate call sites incrementally (Vue Query mutations first, since they already centralize `onError`).
- Fill in the message map as the API audit produces entries; unknown codes keep working via fallback.

This lets the refactor and the audit proceed independently and unblocks #199 from being fully gated on the audit.

cc @joshuagraber @josh-chamberlain

**Comment by maxachis** [comment]
## Backend error audit for #199

Following up on the proposal above — doing the API-side audit so the client refactor and the message map aren't blocked. @joshuagraber, this is meant as a starting draft for you to refine, not a final mapping.

Audit of the two backends consumed by the pdap.io client:

- **`data-sources-app`** — Flask-RESTX (mounted at `/api/v2`) + FastAPI (mounted at `/api/v3`), served at `data-sources.pdap.io`. ASGI composition in `app.py:192-222`.
- **`data-source-manager`** (the "source-collector" backend, `VITE_SM_API_URL`) — pure FastAPI, served at `source-collector.pdap.io`. Confirmed by the `/annotate/anonymous` route in `data-source-manager/src/api/endpoints/annotate/routes.py:40` matching `src/api/annotate.ts:10` in the client.

### 1. How consistent is the error shape?

**Not consistent.** There are at least 4 distinct error body shapes the client may receive across the two backends, plus FastAPI's own validation shape. No global exception handler is registered in either app, so every framework default leaks through:

| # | Shape | Where it comes from | Example body |
|---|-------|---------------------|--------------|
| 1 | `{"message": "..."}` | Flask-RESTX rendering werkzeug `HTTPException` (dominant DSA v2 shape) | `{"message": "User does not have permission to perform this action."}` |
| 2 | `{"detail": "..."}` | FastAPI `HTTPException(detail=str)` (DSA v3 + most of DSM) | `{"detail": "URL not found"}` |
| 3 | `{"detail": {"code": "...", "message": "..."}}` | DSM `ErrorManager.raise_error` — the only place a code is emitted (`data-source-manager/src/core/error_manager/core.py:13-24`) | `{"detail": {"code": "ANNOTATION_EXISTS", "message": "Annotation of type ... already exists for url 42"}}` |
| 4 | `{"detail": {<arbitrary dto>}}` | DSM duplicate-submission path stuffs a Pydantic model into `detail` (`src/api/endpoints/submit/data_source/queries/duplicate.py:53-62`) | `{"detail": {"message": "Duplicate URL found", "url_id": 1, "url_status": "...", "url_type": "..."}}` |
| 5 | `{"detail": [{"loc":[...], "msg":"...", "type":"..."}]}` | FastAPI / Pydantic 422 validation | standard FastAPI |

DSA's Flask-RESTX layer can also return HTML 500 pages on uncaught exceptions (no `errorhandler` is registered anywhere). DSA's column-permission middleware and `dynamic_request_logic` raise werkzeug `Forbidden`/`BadRequest`/`Conflict` directly — all shape #1. Marshmallow under flask-restx produces its own `{"errors":{"field":["msg"]}}` shape — a sixth variant.

The only structured/codified error in the system today is `ErrorTypes.ANNOTATION_EXISTS` in `data-source-manager/src/core/error_manager/enums.py` — a single-member enum.

### 2. Proposed normalized shape

Building on DSM's existing `ErrorFormat`:

```json
{
  "error": {
    "code": "ANNOTATION_EXISTS",
    "message": "Human-readable default (server-supplied)",
    "details": { "url_id": 42 },
    "field_errors": [
      {"field": "agency_info.location_id", "message": "required"}
    ]
  }
}
```

Migration sketch: register one `app.add_exception_handler(HTTPException, ...)` + `RequestValidationError` handler per FastAPI app, and a Flask-RESTX `@api.errorhandler(HTTPException)` under `/api/v2`, each coercing into this envelope. ~50-line change per repo. The client keys UX off `error.code` and falls back to `error.message`.

### 3. Error catalogue (proposed codes)

`code` is **proposed** — only `ANNOTATION_EXISTS` exists today. User-message column intentionally blank for @joshuagraber to fill.

#### Auth / session (both repos)
| code | http | trigger | user message |
|---|---|---|---|
| `INVALID_TOKEN` | 401 | Bad/expired JWT. `data-sources-app/middleware/security/auth/fastapi.py:17`, `data-source-manager/src/security/manager.py:28`. | |
| `INSUFFICIENT_PERMISSIONS` | 403 | Token valid, missing permission. Same files, `:27` and `:58`. | |
| `FORBIDDEN_RESOURCE` | 403 | Row-level deny — `column_permission/core.py:62`, `dynamic_request_logic/delete.py:17`, `dynamic_request_logic/get/many.py:93`, `data_requests_/withdraw.py:21`, `data_requests_/related/related_source/post.py:31`, `endpoints/v3/user/by_id/get/wrapper.py:17`. | |
| `INVALID_API_KEY` | 401 | `middleware/primary_resource_logic/api_key.py:53,65`. | |
| `INVALID_CREDENTIALS` | 401 | Bad email/password — `login_queries.py:67-73` (also `"Email not verified."`). | |
| `OAUTH_TOKEN_EXPIRED` | 401 | GitHub OAuth — `github_oauth.py:105,148`. | |
| `RESET_TOKEN_INVALID` | 400/401 | `reset_token_queries.py:89,104,131,136` (mix of 400 and 401 — see open questions). | |

#### Sign-up / email validation (DSA)
| code | http | trigger |
|---|---|---|
| `USER_EXISTS` | 409 | `endpoints/instantiations/auth_/signup/middleware.py:22-25`, `user_queries.py:61` |
| `EMAIL_NOT_PENDING_VALIDATION` | 400 | `endpoints/instantiations/auth_/resend_validation_email/middleware.py:19` |
| `INVALID_VALIDATION_TOKEN` | 400 | `endpoints/instantiations/auth_/validate_email/query.py:52` |
| `DEV_PASSWORD_INCORRECT` | 401 | `endpoints/instantiations/dev_/route.py:59` |

#### Data sources (DSA)
| code | http | trigger |
|---|---|---|
| `DATA_SOURCE_NOT_FOUND` | 404 | `endpoints/instantiations/data_sources_/get/by_id/wrapper.py:26` |
| `DATA_SOURCE_DUPLICATE` | 409 | `endpoints/instantiations/data_sources_/post/request_/query.py:74-79` |
| `DATA_SOURCE_CREATE_FAILED` | 500 | same file, `:60,82` |

#### Data requests (DSA)
| code | http | trigger |
|---|---|---|
| `REQUEST_NOT_FOUND` | 400 | `data_requests_/related/related_source/post.py:38` (400, not 404) |
| `REQUEST_SOURCE_DUPLICATE` | 409 | same file, `:56` |
| `REQUEST_PERMISSION_DENIED` | 403 | `data_requests_/helpers.py:12`, `data_requests_/withdraw.py:21` |

#### Search / location (DSA)
| code | http | trigger |
|---|---|---|
| `LOCATION_NOT_FOUND` | 400 | `primary_resource_logic/locations.py:23,58`, `search/wrappers/follow/create.py:24` (400, not 404) |
| `INVALID_OUTPUT_FORMAT` | 400 | `search/helpers.py:100` |
| `INVALID_RECORD_CATEGORY_COMBO` | 400 | `search/helpers.py:123` |
| `INVALID_LOCATION` | 500? | `location_logic.py:46,62` — raises `InvalidLocationError`, not caught → bubbles to 500 today |

#### Permissions admin (DSA)
| code | http | trigger |
|---|---|---|
| `USER_NOT_FOUND` | 400 | `permissions.py:19` (400, not 404) |
| `PERMISSION_ALREADY_ASSIGNED` | 409 | `permissions.py:34,41` |

#### Source-collector / annotation (DSM — what `pages/annotate` hits)
| code | http | trigger |
|---|---|---|
| `URL_NOT_FOUND` | 404 | `src/api/endpoints/url/by_id/screenshot/wrapper.py:17-21` (matches `${URL_BASE}/{url_id}/screenshot` in the client); `src/api/shared/check/url_type/query.py:44` |
| `URL_NOT_VALIDATED` | 400 | `src/api/shared/check/url_type/query.py:50` |
| `URL_TYPE_MISMATCH` | 400 | `src/api/shared/check/url_type/query.py:55` |
| `ANNOTATION_EXISTS` | 409 | **only enum-coded error.** `src/core/error_manager/core.py:26-36` — wrapped shape #3 |
| `ANNOTATION_VALIDATION` | 400 | `FailedValidationException` in `src/api/endpoints/annotate/all/post/models/request.py:28-40`, `src/core/tasks/url/operators/validate/queries/get/models/response.py:21-65` |
| `BATCH_NOT_FOUND` | 400 | `src/core/core.py:63` (400, not 404) |
| `AGENCY_NOT_FOUND` | 400 | `src/api/endpoints/agencies/by_id/put/query.py:34`, `src/api/endpoints/review/approve/query_/core.py:151`, `src/api/endpoints/proposals/agencies/by_id/put/query.py:34` |
| `AGENCY_REQUIRED` | 400 | `src/api/endpoints/review/approve/query_/core.py:124` |
| `REVIEW_REJECT_INVALID` | 400 | `src/api/endpoints/review/reject/query.py:46` |
| `SEARCH_INVALID` | 400 | `src/api/endpoints/search/routes.py:51` |
| `DUPLICATE_SUBMISSION` | 409 | `src/api/endpoints/submit/data_source/queries/duplicate.py:59` — uses shape #4 (DTO inside `detail`) |
| `SUBMIT_DATA_SOURCE_BAD_REQUEST` | 400 | `src/api/endpoints/submit/data_source/wrapper.py:21` |
| `AGENCY_CREATE_FAILED` | 500 | `src/api/shared/agency/post/query.py:29` |
| `TASK_NOT_FOUND` | **200 (sic)** | `src/collectors/manager.py:71` raises `HTTPException(status_code=HTTPStatus.OK, ...)` — almost certainly a bug |

#### Validation (both repos)
Standard FastAPI 422 `{"detail":[{"loc":[...], "msg":"...", "type":"..."}]}`. DSM has many `ValueError` raises in Pydantic validators that surface this way. DSA's flask-restx layer uses marshmallow → `{"errors":{"field":["msg"]}}` (shape #6).

### 4. Open questions for @joshuagraber

1. **Status-code conventions.** Many "X not found" cases return **400 instead of 404** (locations, permissions/users, data requests, batches, agencies in DSM). Intentional, or normalize to 404?
2. **`TASK_NOT_FOUND` returning HTTP 200** — bug? (`data-source-manager/src/collectors/manager.py:71`)
3. **Uncaught application exceptions** — `InvalidLocationError`, `MuckrockAPIError`, `MatchAgencyError`, `CKANAPIError`, `IntegrityMonitorTaskException`, `SubtaskError` are raised but never translated. They surface as 500 with framework-default body (HTML in Flask, generic in FastAPI). Should these get codes?
4. **Validation shapes** — unify marshmallow / Pydantic / werkzeug into one, or accept three?
5. **Auth granularity** — today every "missing permission" is `403 "Access forbidden"` and every "bad token" is `401 "Could not validate credentials"`. The client cannot distinguish *expired* from *malformed* from *revoked* tokens. Worth splitting?
6. **Naming**: `SCREAMING_SNAKE` (matches existing `ANNOTATION_EXISTS`) or dotted (`annotation.exists`)?
7. **Reset-token 400 vs 401** — `reset_token_queries.py` mixes both for conceptually similar cases. Pick one?

### 5. Recommended sequencing

1. Client ships the parsing architecture (layers 1/2/4 from the previous comment) with a passthrough map — defensively reads both `error.detail` and `error.message` and handles object-with-`code` specially. No backend change required.
2. Each backend adds a single exception handler coercing to the proposed envelope. ~50 lines per repo.
3. Fill in the `code → userMessage` map from the table above, with @joshuagraber's copy.
4. Fix the status-code bugs (open questions 1, 2) in a separate PR.

cc @maxachis @josh-chamberlain
